src/cdk-experimental/ui-patterns/listbox/listbox.ts

```diff
@@ -130,40 +144,74 @@ export class ListboxPattern<V> {
     }
 
     if (this.inputs.multi()) {
+      // The following block implements range selection using a keyboard.
+      //
+      // Range selection is started by holding shift and navigating to the next item in a list.
+      //
+      // Wrapping is disabled during range selection since any decision on how to
+      // continue the range after wrapping would lead to an unintuitive user experience.
+      manager
+        .on(Modifier.Any, 'Shift', () => this.anchorIndex.set(this.inputs.activeIndex()))
+        .on(Modifier.Shift, this.prevKey, () => {
+          if (this.inputs.activeIndex() === this.anchorIndex()) {
+            this.selection.beginRangeSelection();
+          }
+          this.wrap.set(false);
+          this.prev({selectRange: true});
+          this.wrap.set(true);
+        })
+        .on(Modifier.Shift, this.nextKey, () => {
+          if (this.inputs.activeIndex() === this.anchorIndex()) {
+            this.selection.beginRangeSelection();
+          }
+          this.wrap.set(false);
+          this.next({selectRange: true});
+          this.wrap.set(true);
+        });
+
       manager
-        .on(Modifier.Shift, 'Enter', () => this._updateSelection({selectFromAnchor: true}))
-        .on(Modifier.Shift, this.prevKey, () => this.prev({toggle: true}))
-        .on(Modifier.Shift, this.nextKey, () => this.next({toggle: true}))
-        .on([Modifier.Ctrl, Modifier.Meta], 'A', () => this._updateSelection({selectAll: true}))
-        .on([Modifier.Ctrl | Modifier.Shift, Modifier.Meta | Modifier.Shift], 'Home', () =>
-          this.first({selectFromActive: true}),
-        )
-        .on([Modifier.Ctrl | Modifier.Shift, Modifier.Meta | Modifier.Shift], 'End', () =>
-          this.last({selectFromActive: true}),
-        )
-        .on(Modifier.Shift, this.dynamicSpaceKey, () =>
-          this._updateSelection({selectFromAnchor: true}),
-        );
+        .on(Modifier.Shift, 'Enter', () => {
+          this.selection.selectRange();
+          this.anchorIndex.set(this.selection.rangeStartIndex());
+        })
+        .on(Modifier.Shift, this.dynamicSpaceKey, () => {
+          this.selection.selectRange();
+          this.anchorIndex.set(this.selection.rangeStartIndex());
+        })
+        .on([Modifier.Ctrl | Modifier.Shift, Modifier.Meta | Modifier.Shift], 'Home', () => {
```

Comment:
I wonder if you could add a shortcut to the `Modifier` enum like you did for `All`: `Modifer.CtrlOrMeta`. Trying to figure out how it plays with bit logic might be weird though, not sure. It seems like a common enough case to maybe be worth considering.

Or, just bake into `KeyboardEventManager` logic to normalize whatever modifiers it gets across platforms. So it would normalize `Ctrl | Shift`, `Meta | Shift`, and `[Ctrl | Shift, Meta | Shift]` to the same thing: `[Ctrl | Shift, Meta | Shift]`

Reply:
I tried adding this to the EventManager but I couldn't get the bit logic to play nicely - specifically when trying to join `CtrlMeta` with other modifiers. I'm hesitant to bake them into the `KeyboardEventManager` for now because I'm not certain that all `Ctrl` & `Meta` events are meant to get this treatment